### ISSUE TITLE
Preserve image metadata (EXIF, IPTC) from source image when generating target image 

### ISSUE THREAD

**Issue by schwemmer** [loris-imageserver/loris#126]
Would it be possible to preserve the image metadata such as EXIF and IPTC in the generated image? 
The attached JPEG is a sample image that contains some metadata (normally we use JP2 files but I can't upload one here). If you open it in Photoshop or another application that can read image metadata you will find some metadata such as contact information, copyright notice, camera settings etc. (see attached screenshots). In the generated target image this metadata is lost. It would be great if we could somehow preserve it.

![bbb-0009_011v](https://cloud.githubusercontent.com/assets/725398/4310776/4e719000-3ea9-11e4-9408-c60d952e154b.jpg)

![screenshot 2014-09-17 22 30 18](https://cloud.githubusercontent.com/assets/725398/4310822/a1b0538c-3ea9-11e4-8137-fb76bbb72508.png)

![screenshot 2014-09-17 22 31 21](https://cloud.githubusercontent.com/assets/725398/4310827/a83e3ab6-3ea9-11e4-801e-aa003ae47344.png)

**Comment by jpstroop** [comment]
The info from your file (thanks for that!) is below. My concern is that some of it is logical (e.g. Exif.Image.Artist, Exif.Image.Copyright, the IPTC data, and then the XMP representations of those) and therefore makes sense for copying over; while most of the exif fields are technical and aren't necessarily true about the downstream derivative file (e.g. dimensions, orientation, bits per sample); and some falls somewhere in between, like camera/capture data.

To avoid too much philosophizing (e.g. should the camera data really come over for a file that is likely at least two derivations away from the file that was originally captured?, should the software be Loris? Pillow? Kakadu?) and configuration nightmares, I think the cleanest solution might be to stick to copying over the IPTC data only. Furthermore, because there would be a performance hit, and extra dependencies would be required, I'd probably make this feature configurable--on or off--and off by default.

What do you think?

| Tag | Value |
| :-- | :-- |
| Exif.Image.Artist | CESG-Repro 2 |
| Exif.Image.BitsPerSample | 8 8 8 |
| Exif.Image.Compression | 1 |
| Exif.Image.Copyright | Burgerbibliothek Bern / e-codices |
| Exif.Image.DateTime | 2011:11:03 14:48:55 |
| Exif.Image.ExifTag | 380 |
| Exif.Image.ImageLength | 6496 |
| Exif.Image.ImageWidth | 4872 |
| Exif.Image.Make | Hasselblad |
| Exif.Image.Model | Hasselblad H3DII-31 |
| Exif.Image.Orientation | 1 |
| Exif.Image.PhotometricInterpretation | 2 |
| Exif.Image.PlanarConfiguration | 1 |
| Exif.Image.ResolutionUnit | 2 |
| Exif.Image.SamplesPerPixel | 3 |
| Exif.Image.Software | Adobe Photoshop CS3 Windows |
| Exif.Image.XResolution | 3000000/10000 |
| Exif.Image.YResolution | 3000000/10000 |
| Exif.Photo.ApertureValue | 536870912/67108864 |
| Exif.Photo.ColorSpace | 1 |
| Exif.Photo.DateTimeDigitized | 2011:09:19 12:19:03 |
| Exif.Photo.DateTimeOriginal | 2011:09:19 12:19:03 |
| Exif.Photo.ExifVersion | 48 50 49 48 |
| Exif.Photo.ExposureProgram | 1 |
| Exif.Photo.ExposureTime | 2147483/536870912 |
| Exif.Photo.FNumber | 536870912/33554432 |
| Exif.Photo.Flash | 0 |
| Exif.Photo.FocalLength | 671088640/8388608 |
| Exif.Photo.ISOSpeedRatings | 100 |
| Exif.Photo.MeteringMode | 3 |
| Exif.Photo.PixelXDimension | 609 |
| Exif.Photo.PixelYDimension | 812 |
| Exif.Photo.ShutterSpeedValue | 1069149468/134217728 |
| Exif.Thumbnail.Compression | 6 |
| Exif.Thumbnail.JPEGInterchangeFormat | 742 |
| Exif.Thumbnail.JPEGInterchangeFormatLength | 5883 |
| Exif.Thumbnail.ResolutionUnit | 2 |
| Exif.Thumbnail.XResolution | 72/1 |
| Exif.Thumbnail.YResolution | 72/1 |
| Iptc.Application2.0x0038 | 121903 |
| Iptc.Application2.Byline | CESG-Repro 2 |
| Iptc.Application2.BylineTitle | Fotograf/innen |
| Iptc.Application2.Copyright | Burgerbibliothek Bern / e-codices |
| Iptc.Application2.Credit | e-codices |
| Iptc.Application2.DateCreated | 2011-09-19 |
| Iptc.Application2.RecordVersion | 4 |
| Xmp.dc.creator | CESG-Repro 2 |
| Xmp.dc.format | image/jpeg |
| Xmp.dc.rights | lang="x-default" Burgerbibliothek Bern / e-codices |
| Xmp.exif.ApertureValue | 536870912/67108864 |
| Xmp.exif.ColorSpace | 1 |
| Xmp.exif.DateTimeDigitized | 2011-09-19T12:19:03+02:00 |
| Xmp.exif.DateTimeOriginal | 2011-09-19T12:19:03+02:00 |
| Xmp.exif.ExifVersion | 0210 |
| Xmp.exif.ExposureProgram | 1 |
| Xmp.exif.ExposureTime | 2147483/536870912 |
| Xmp.exif.FNumber | 536870912/33554432 |
| Xmp.exif.Flash/exif:Fired | False |
| Xmp.exif.Flash/exif:Function | False |
| Xmp.exif.Flash/exif:Mode | 0 |
| Xmp.exif.Flash/exif:RedEyeMode | False |
| Xmp.exif.Flash/exif:Return | 0 |
| Xmp.exif.FocalLength | 671088640/8388608 |
| Xmp.exif.ISOSpeedRatings | 100 |
| Xmp.exif.MeteringMode | 3 |
| Xmp.exif.NativeDigest | 36864,40960,40961,37121,37122,40962,40963,37510,40964,36867,36868,33434,33437,34850,34852,34855,34856,37377,37378,37379,37380,37381,37382,37383,37384,37385,37386,37396,41483,41484,41486,41487,41488,41492,41493,41495,41728,41729,41730,41985,41986,41987,41988,41989,41990,41991,41992,41993,41994,41995,41996,42016,0,2,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,20,22,23,24,25,26,27,28,30;DE0A0560F94861F8F0B4EE5012A9CE03 |
| Xmp.exif.PixelXDimension | 609 |
| Xmp.exif.PixelYDimension | 812 |
| Xmp.exif.ShutterSpeedValue | 1069149468/134217728 |
| Xmp.iptc.CreatorContactInfo/Iptc4xmpCore:CiAdrCity | St. Gallen |
| Xmp.iptc.CreatorContactInfo/Iptc4xmpCore:CiAdrCtry | Schweiz |
| Xmp.iptc.CreatorContactInfo/Iptc4xmpCore:CiAdrExtadr | Klosterhof 6d |
| Xmp.iptc.CreatorContactInfo/Iptc4xmpCore:CiAdrPcode | 9004 |
| Xmp.iptc.CreatorContactInfo/Iptc4xmpCore:CiTelWork | +41 71 227 34 31 |
| Xmp.iptc.CreatorContactInfo/Iptc4xmpCore:CiUrlWork | www.e-codices.unifr.ch |
| Xmp.photoshop.AuthorsPosition | Fotograf/innen |
| Xmp.photoshop.ColorMode | 3 |
| Xmp.photoshop.Credit | e-codices |
| Xmp.photoshop.DateCreated | 2011-09-19 |
| Xmp.photoshop.ICCProfile | sRGB IEC61966-2.1 |
| Xmp.tiff.BitsPerSample | 8, 8, 8 |
| Xmp.tiff.Compression | 1 |
| Xmp.tiff.ImageLength | 6496 |
| Xmp.tiff.ImageWidth | 4872 |
| Xmp.tiff.Make | Hasselblad |
| Xmp.tiff.Model | Hasselblad H3DII-31 |
| Xmp.tiff.NativeDigest | 256,257,258,259,262,274,277,284,530,531,282,283,296,301,318,319,529,532,306,270,271,272,305,315,33432;BEAAB8A9748BA4067DFA3256CFB7B451 |
| Xmp.tiff.Orientation | 1 |
| Xmp.tiff.PhotometricInterpretation | 2 |
| Xmp.tiff.PlanarConfiguration | 1 |
| Xmp.tiff.ResolutionUnit | 2 |
| Xmp.tiff.SamplesPerPixel | 3 |
| Xmp.tiff.XResolution | 3000000/10000 |
| Xmp.tiff.YResolution | 3000000/10000 |
| Xmp.xmp.CreateDate | 2011-11-03T14:48:55+01:00 |
| Xmp.xmp.CreatorTool | Adobe Photoshop CS3 Windows |
| Xmp.xmp.MetadataDate | 2011-11-03T14:48:55+01:00 |
| Xmp.xmp.ModifyDate | 2011-11-03T14:48:55+01:00 |
| Xmp.xmpMM.DerivedFrom/stRef:documentID | uuid:C7E31F922206E1119803AA1F25C80FF3 |
| Xmp.xmpMM.DerivedFrom/stRef:instanceID | uuid:C8E31F922206E1119803AA1F25C80FF3 |
| Xmp.xmpMM.DocumentID | uuid:CBE31F922206E1119803AA1F25C80FF3 |
| Xmp.xmpMM.InstanceID | uuid:CCE31F922206E1119803AA1F25C80FF3 |
| Xmp.xmpRights.UsageTerms | lang="x-default" http//www.e-codices.unifr.ch/en/info/terms |

**Comment by jpstroop** [comment]
E.g.

```
sudo apt-get install libexiv2-dev
sudo apt-get install  gir1.2-gexiv2-0.10
```

``` python
>>> from gi.repository import GExiv2
>>> from sys import stdout
>>> md = GExiv2.Metadata('tests/img/ecodicies.jpg')
>>> tags = filter(lambda t: 'iptc' in t, md.get_xmp_tags()) + md.get_iptc_tags()
>>> [stdout.write('%s: %s\n' % (t, md[t])) for t in tags]
```

```
Xmp.iptc.CreatorContactInfo/Iptc4xmpCore:CiAdrCity: St. Gallen
Xmp.iptc.CreatorContactInfo/Iptc4xmpCore:CiAdrCtry: Schweiz
Xmp.iptc.CreatorContactInfo/Iptc4xmpCore:CiAdrExtadr: Klosterhof 6d
Xmp.iptc.CreatorContactInfo/Iptc4xmpCore:CiAdrPcode: 9004
Xmp.iptc.CreatorContactInfo/Iptc4xmpCore:CiTelWork: +41 71 227 34 31
Xmp.iptc.CreatorContactInfo/Iptc4xmpCore:CiUrlWork: www.e-codices.unifr.ch
Iptc.Application2.0x0038: 121903
Iptc.Application2.Byline: CESG-Repro 2
Iptc.Application2.BylineTitle: Fotograf/innen
Iptc.Application2.Copyright: Burgerbibliothek Bern / e-codices
Iptc.Application2.Credit: e-codices
Iptc.Application2.DateCreated: 2011-09-19
Iptc.Application2.RecordVersion: 4
```

**Comment by jpstroop** [comment]
OK, check out the branch referenced above; it still needs some tidying up, but assuming you're on an Ubuntu system, do:

```
$ sudo apt-get install libexiv2-dev gir1.2-gexiv2-0.10
```

(Exact version of gir1.2-gexiv2-0.10 may vary).

and set `copy_iptc = True` in the config file.

I haven't figured out how to get gexiv2 to work with virtualenv yet; I'm sure there must be a way, but if that's a concern for you you may need to make some symlinks or something.

I'd be curious to hear if it's working with your JP2s in particular.

**Comment by schwemmer** [comment]
Thank you so much for your extremely quick response! I agree with you that the Exif data is problematic. Our idea, originally, was to display some info about exposure time, f-stop etc. alongside the image, similar to what Flickr does. However, you are right that this is questionable for an image that is several derivatives away from the original image (in our case: RAW => TIFF => JP2 => JPEG) and for some metadata fields plain wrong (e.g. dimensions, when requesting a smaller version). 

As for the Exif fields that do make sense, namely Exif.Image.copyright, we have those available in IPTC too. This is the most important metadata field for us, so your solution to only copy over the IPTC data will work for us. I agree that this should be configurable and off by default as performance has higher priority and not all people will be using this feature. In short, your proposed solution seems perfect, I will now try to get it working on our server (Debian 6) and provide some more feedback after testing it.

**Comment by schwemmer** [comment]
Hi Jon, I tried to install the required dependencies on my server (Debian 6). The libexiv2-dev package was no problem, but there is no gir1.2-gexiv2-0.10 or equivalent/similar package available for Debian 6. After a few hours of trying I gave up and asked @regisrob if he has an idea on how to install the dependency on Debian 6. While he also wasn't able to install the required package on Debian he did find a possible fallback solution. Here is what he wrote:

> The only alternative seems to be the old Python binding to Exiv2 called 
> "python-pyexiv2": http://tilloy.net/dev/pyexiv2/
> It is deprecated but it works from the command-line:
> 
> ```
> # sudo aptitude install python-pyexiv2
> # python
> >>> import pyexiv2
> >>> md = pyexiv2.Image('/path/to/image/image.jpg')
> >>> md.readMetadata()
> >>> print md.iptcKeys()
> >>> tag = md['Iptc.Application2.Copyright']
> >>> print tag
> Burgerbibliothek Bern / e-codices
> ```

I tried this on my server and it worked from the command line. Régis suggested that maybe you could add some kind of fallback to your code along the lines of "if 'from gi.repository import GExiv2' raises an ImportError, then check if 'pyexiv2' is available and loop through the embedded metadata with the methods provided by this module). 
What do you think?

**Comment by jpstroop** [comment]
I don't know Debian at all, but I wonder if there's a way of getting backports? I see this:

https://packages.debian.org/source/sid/gexiv2

but assume that's for Debian 7?

otherwise, yeah, I guess it's pyexiv2 (which I was hoping to avoid).

Is building from source (:crying_cat_face: I realize) and option?

https://wiki.gnome.org/action/show/Projects/gexiv2

**Comment by jpstroop** [comment]
OK...building from source is not an option! I just spun up a Debian Squeeze box from scratch and quickly wound up in dependency hell (or purgatory....not sure which seems more appropriate :imp: ). So it looks like the pyexiv2 might be the way to go, at least for a while.

I'll try to have a look in the next day (evening) or two.

**Comment by jpstroop** [comment]
Hmm...the console code you pasted above doesn't work for me. What does this:

``` python
>>> pyexiv2.version_info
(0, 3, 2)
```

return for you?

**Comment by schwemmer** [comment]
Sorry for the trouble. Yes, dependency hell indeed, I've been there too. Strangely, the code from Régis works for me but when I do 

```
>>> pyexiv2.version_info
```

all I get is an error:

```
AttributeError: 'module' object has no attribute 'version_info'
```

even though it's in the docs: http://tilloy.net/dev/pyexiv2/doc/release-0.3.0/api.html

To spare you too much trouble with this I can consider upgrading to Debian 7 – not sure if it's feasible though. I will have to talk to the IT department first. It might be the easier way to go, unless other people need Debian 6 support too.

**Comment by jpstroop** [comment]
urgh...looks like version 0.1.3 is what apt has for debian 6, and that method doesn't even exist!

**Comment by jpstroop** [comment]
As far as I know the debian users are you and @regisrob, but if I can handle it the differences, I don't mind trying. Just have to try to guess the older API...unless either of you know how to _write_ tags with version 0.1.3.

**Comment by jpstroop** [comment]
Found the docs. Retro!

http://tilloy.net/dev/pyexiv2/doc/release-0.1.3/pyexiv2.htm

**Comment by jpstroop** [comment]
Hmm...I think you might want to consider an upgrade. 0.1.3 doesn't support XMP :frowning:.

**Comment by schwemmer** [comment]
Wow, these docs look _really_ old :-) Thanks for trying! I appreciate it. I think upgrading to Debian 7 makes sense, at least if we want this feature. I will see what I can do tomorrow and report back.

**Comment by schwemmer** [comment]
Hi Jon, I upgraded to Debian 7 without problems today. However, my Loris server stopped working and I spent all day trying to get it back to working but without success. I am getting "undefined symbol: PyUnicodeUCS4_AsLatin1String" error messages and 500 errors.

Also, I just noticed now that the required gir1.2-gexiv2-0.10 package isn't even available on Debian 7 (Wheezy) but only on the next unreleased Debian version (Jessie): https://packages.debian.org/search?keywords=gir1.2-gexiv2-0.10

So probably we will have to forgo using the image metadata feature for now and I might go back to Debian 6 (luckily we did a snapshot before upgrading) where my Loris server is at least working.

**Comment by jpstroop** [comment]
Youch, I'm sorry!

If you still have the machine, can you check the version of python? I can't help but wonder if you were running with a 3.x.y version of python.

**Comment by schwemmer** [comment]
After all, we were able to solve the problem now. We were using Python 2.7.3 that comes with Debian 7. The first problem was that _functools was not found which was fixed by adding this to the WSGI file:

```
import sys
sys.path.append('/usr/local/lib/python2.7/lib-dynload') 
sys.path.append('/usr/local/lib/python2.7/site-packages')
sys.path.append('/usr/local/lib/python2.7/dist-packages')
```

The second problem, the "undefined symbol: PyUnicodeUCS4_AsLatin1String" error was fixed by using a custom version of Python compiled with Unicode support like this:

```
% wget http://www.python.org/ftp/python/2.7.3/Python-2.7.3.tgz
% tar -xzvf Python-2.7.3.tgz
% cd Python-2.7.3
% ./configure --enable-shared --with-threads --enable-unicode=ucs4  <--- THIS IS THE IMPORTANT OPTION
% make
% sudo make altinstall
```

So now it's all working on Debian 7, great! The next step will be to get the gir1.2-gexiv2-0.10 package or something equivalent working on Debian 7 to use the Loris "exif" branch.

**Comment by jpstroop** [comment]
What a palava! (as my English wife would say). I do I have a version 
that works with pyexiv2 as well, but it doesn't like one of your IPTC 
tags. I'll push it to that same branch when I get into the office.

We also still need to test with one of your JP2s, since--I 
believe--baseline JP2 doesn't actually support exif, iptc, etc. 
(Photoshop produces jpx or jpf, or one of those, which does).

**Comment by jpstroop** [comment]
The error is:

```
NotImplementedError: XMP conversion for type [ContactInfo]
```

I just pushed a branch from the `exif` branch, `exif_with_pyexiv2` in case you want to play around with it, but I think we should try as hard a possible to use gexiv2. Agree?

**Comment by schwemmer** [comment]
I agree that Gexiv2 will be the way to go. We did try as hard as possible to install it on Debian 7 but it's dependency hell all over again. We tried installing it from experimental Debian sources, from the Jessie (next Debian release) sources, downloading it from Github and compiling it etc. The best we could get is that Python finds the module but complains about "typelib missing" and everyone on the internet having the same error said to install gir1.2-gexiv2-0.x which we tried in the first place but it's not available on Wheezy and doesn't work because of dependency problems when downloaded from Jessie. In short, I don't think it's possible at all and we will have to wait until Jessie becomes stable and we can upgrade the system.

We also tried the exif_with_pyexiv2 branch (thanks for that!). We got it working but as you said, the ContactInfo tag is not implemented in pyevix2 (apparently this is a known "bug" since beginning of 2012: https://bugs.launchpad.net/pyexiv2/+bug/914598, so I don't think it will ever be fixed). Removing the ContactInfo from our image metadata is not really an option, unless it could be done programmatically somehow (ca. 425'000 images). Also, when I removed the ContactInfo from one of the images as a test we got different errors (NameError: global name 'copy' is not defined).

To sum up, I think we will have to defer using this feature until we can upgrade to the next Debian version as soon as it becomes stable. The way to go is Gexiv2 and this should be no problem to install on the next Debian version.

**Comment by jpstroop** [comment]
If that's OK with you, I think you're correct. I'll polish the feature (the `exif` branch) and get it in sometime soon so it'll be there when you're ready for it. Thanks for your efforts though, and in the meantime if another approach reveals itself, I'm happy to explore it.

None of my business, of course, but you could always "upgrade" to Ubuntu. :wink:

**Comment by schwemmer** [comment]
Yes, this sounds like a good plan. Thanks again very much for implementing this so quickly!! I might look into "upgrading" to Ubuntu later but I don't know if the university even offers this option for their virtual servers and it would probably be a lot of work to set up everything from scratch (Apache, Solr, etc. etc.). For the time being I guess we'll have to wait a little before using the image metadata feature. The next Debian release should "freeze" in November and probably become stable sometime in spring next year.

**Comment by jpstroop** [comment]
> I might look into "upgrading" to Ubuntu later

I hope you understand that I'm kidding--it's really none of my business and _ideally_ all of this would just work, at least across the various Linux distros. Reality is different, of course.

**Comment by schwemmer** [comment]
Yes, I know that you were kidding ;-) but you've got a point... If our server was running on Ubuntu it would have saved me many hours of work and it seems that Ubuntu has more options in terms of available packages, probably due to its different philosophy (although Debian is said to be slightly more stable and performant). My "dream" would be that one day we could just do:

```
sudo apt-get install loris
```

:relaxed: :sunglasses:

**Comment by jpstroop** [comment]
I know, I know...you're not the first person to ask that. :smiley: In many, many ways it would probably be easier than the `setup.py` approach. Kakadu is always going to make that tricky (because it's proprietary), but I suppose OpenJPEG could be the default. It's just that then new users would be likely to think that Loris' performance is terrible.

**Comment by jpstroop** [comment]
Also, for whatever it's worth, we've been running Ubuntu LTS releases in production since '08 and I can't say we've ever had a problem that we've traced back to Ubuntu or which has made us regret the migration from RHEL (generally quite the opposite, in fact!).

**Comment by schwemmer** [comment]
Interesting. So Ubuntu would really be an option for us too (except for having to set up everything from scratch again for which we probably won't have the time right now). Maybe I'll talk to the IT department later and see what they think.

I guess Debian does not allow proprietary software in their packages (i.e. Kakadu), what about Ubuntu? Or is the problem on Kakadu's side, that they don't allow it? But it's okay to bundle it?

Performance is more important IMHO than ease of installation so I would prefer the current solution over OpenJPEG by default.

By the way, the supplied Kakadu 7.2 doesn't work on Debian (neither Debian 6 nor 7) due to glibc version incompatibilities and the paths to Kakadu in the loris config file seem to be ignored by Loris. I have this in my loris config file:

```
kdu_expand = '/usr/local/bin/kdu_expand' # r-x
kdu_libs = '/usr/local/lib' # r--
```

But I have to actually copy the Kakadu executable and library (v 7.4 in my case) into the loris directory like this to make it work:

```
cp /usr/local/bin/kdu_expand /var/www/loris/bin/Linux/x86_64/kdu_expand
cp /usr/local/lib/libkdu_v74R.so /var/www/loris/lib/Linux/x86_64/
```

Would you like me to file a separate issue for this?

**Comment by jpstroop** [comment]
> I guess Debian does not allow proprietary software in their packages (i.e. Kakadu), what about Ubuntu? Or is the problem on Kakadu's side, that they don't allow it? But it's okay to bundle it?

Ubuntu is based on Debian, and uses APT as well, and as far as I can tell the conventions in terms of where things live on the filesystem are pretty similar (when I was playing with Debian 6 last week it felt very similar to Ubuntu, just a little older, and I didn't really have any trouble finding my way around). I don't know the exact details in terms of what they allow and don't allow in APT, but, regardless, I doubt Kakadu would go for it. They have some fuzzy licensing terms that allow you to evaluate the software and use the binaries they distribute in applications that aren't for profit, but it's definitely not OSS, and they generally want you to pay. They're far from legal experts though, and the exact terms of usage seem to be based on who answers your query (I know a number of people who have asked the same questions as me and we've all received a variety of answers). The correspondence I had with them is [in this repo](https://github.com/pulibrary/loris/blob/development/doc/kdu_license_inquiry.eml#L44-L165) for safe keeping.

> Performance is more important IMHO than ease of installation so I would prefer the current solution over OpenJPEG by default.

That's definitely my philosophy too, and all I can say is that installation gets easier with every release, and hopefully the docs help. If you ask Regis, I think he'll tell you that it used to be _much_ worse. :smile: 

> By the way, the supplied Kakadu 7.2 doesn't work on Debian (neither Debian 6 nor 7) due to glibc version 

I filed #109 a while ago, I think after a conversation about it w/Regis, and he followed up on the ticket. It's just something I need to get to...or am waiting for someone else to do. :wink: I'd like to do a release before London, and I think this would be included in it.

**Comment by schwemmer** [comment]
Thanks for these clarifications! I will comment on #109 regarding the Kakadu executables. 

As for installation, I am glad that @regisrob wrote a guide for Debian (http://doc.biblissima-condorcet.fr/loris-setup-guide-ubuntu-debian) and I can imagine that it used to be more difficult and that it's getting easier with every release. Cool! Maybe I can contribute at least a little bit to the docs. I'll PM you.

**Comment by deutsches-museum-digital** [comment]
This feature would be important for us too. Now I'm curious about what has become of it? Has it been incorporated into Loris?

As suggested above, I installed the additionally needed software on the Ubuntu 14.04.5 LTS box where Loris is running on:

    sudo apt-get install libexiv2-dev gir1.2-gexiv2-0.10

And configured Loris like so in /etc/loris2/loris2.conf:

    [transforms]
        ...
        copy_iptc = True

And restarted the web server. But these steps hadn't got the desired effect. IPTC data that definitively exist in an original JPG file are missing in the derivatives that Loris delivers.

I searched the downloaded code for it as well as the Github repository, but could find nothing related besides this page here and the branch mentioned above.

So am I missing something?

**Comment by jpstroop** [comment]
@deutsches-museum-digital The branch that included this feature was never merged. It's been a while, but if I recall correctly, the problem was that we couldn't find a library that would handle extra, custom EXIF fields. I think it's explained above: https://github.com/loris-imageserver/loris/issues/126#issuecomment-56725221

If that's been fixed, or there's another approach that would work and is sustainable to deploy, I'd be happy to see this revisited.

**Comment by deutsches-museum-digital** [comment]
Hmmm ... but in the comments above it says that everybody was happy and content with the feature and that the only problem was that gexiv2 could not be installed in Debian Wheezy. The latter is now oldstable and I can confirm that gexiv2 installs without problems in Debian Jessie (8.2 and 8.7) as well as in Ubuntu 14.04.5 LTS.

Maybe it's time now to think about a merge again? ;)

**Comment by azaroth42** [comment]
We (J Paul Getty Trust) also require this functionality, and are happy to put effort towards ensuring the implementation is production quality.

**Comment by azaroth42** [comment]
Looking at the implementation, could the embedded data be extracted and managed as an external sidecar file (perhaps in JSON) to avoid having to read the full image every time? Check that the modified timestamp of the JSON is > than the modified timestamp of the image, and if not, then re-extract.  This would also allow data to be embedded by the server, regardless of where it came from (manually created, automated from a DAMs, embedded IPTC, etc)

**Comment by rogerhoward** [comment]
@azaroth42 some things I considered as I tackled similar questions (I use IIIF as a fun challenge and teaching aid, and have implemented it myself in a number of langs/stacks as a coding kata):

* metadata can quickly become a non-trivial percentage of the total byte size of a smallish image - for instance, the XMP alone in Getty's open content Irises is roughly 5KB, and that doesn't count IPTC/IIM or EXIF. For a 20-50KB tile, that's a substantial amount of bloat.
* I would favor preserving metadata by default, but you'd need to add some optional parameter to the IIIF spec (or by convention) that viewers like OpenSeaDragon can include in their URL patterns to disable the metadata preservation. I handled this previously with a simple query parameter to avoid monkeying with the IIIF URL spec.
* Caching metadata is appropriate - Loris already has a backend cache so this should be straightforward
* I found it handy to add a 'metadata' property to the Image Information response, which added the embedded metadata in a JSON-encoded structure that used exiftool's output with the following switches (from memory, but I think this is right): '-S -g -j -struct'
